rslib/src/notes/mod.rs

```diff
@@ -333,21 +337,52 @@ impl Collection {
         Ok(())
     }
 
-    pub(crate) fn add_note_inner(
+    #[allow(clippy::map_entry)]
+    pub(crate) fn add_notes_inner(
         &mut self,
-        ctx: &CardGenContext<&Notetype>,
-        note: &mut Note,
+        notes: &mut [Note],
         did: DeckId,
         normalize_text: bool,
     ) -> Result<()> {
-        self.canonify_note_tags(note, ctx.usn)?;
-        note.prepare_for_update(ctx.notetype, normalize_text)?;
-        note.set_modified(ctx.usn);
-        self.add_note_only_undoable(note)?;
-        self.generate_cards_for_new_note(ctx, note, did)?;
-        self.set_last_deck_for_notetype(note.notetype_id, did)?;
-        self.set_last_notetype_for_deck(did, note.notetype_id)?;
-        self.set_current_notetype_id(note.notetype_id)
+        let mut notetypes: HashMap<NotetypeId, Arc<Notetype>> = HashMap::new();
+        for note in notes.iter() {
+            if !notetypes.contains_key(&note.notetype_id) {
+                let nt = self
+                    .get_notetype(note.notetype_id)?
+                    .or_invalid("missing note type")?;
+                notetypes.insert(note.notetype_id, nt);
+            }
+        }
+        let mut contexts: HashMap<NotetypeId, CardGenContext<&Notetype>> = HashMap::new();
+        for note in notes.iter_mut() {
+            let last_deck = self.get_last_deck_added_to_for_notetype(note.notetype_id);
+            if !contexts.contains_key(&note.notetype_id) {
+                let ctx = CardGenContext::new(
+                    notetypes.get(&note.notetype_id).unwrap().as_ref(),
+                    last_deck,
+                    self.usn()?,
+                );
+                contexts.insert(note.notetype_id, ctx);
+            }
+        }
+        for note in notes.iter_mut() {
+            let ctx = contexts.get(&note.notetype_id).unwrap();
+            self.canonify_note_tags(note, ctx.usn)?;
+            note.prepare_for_update(ctx.notetype, normalize_text)?;
+            note.set_modified(ctx.usn);
+            self.add_note_only_undoable(note)?;
+            let ctx = contexts.get(&note.notetype_id).unwrap();
+            self.generate_cards_for_new_note(ctx, note, did)?;
+        }
+        for notetype_id in notetypes.into_keys() {
+            self.set_last_deck_for_notetype(notetype_id, did)?;
+            self.set_last_notetype_for_deck(did, notetype_id)?;
+        }
+        if let Some(last_note) = notes.last() {
+            self.set_current_notetype_id(last_note.notetype_id)?;
+        }
```

Comment:
These are quite cheap too. It's great that you're aiming to be efficient here, but in this particular case I think there probably wouldn't be a big difference between this code and just doing the equivalent of `for note_and_deck_id in note_and_deck_ids { old_code() }`. But happy to be proved wrong with a benchmark :-)

Reply:
Yes. I did a quick test and there's no big difference (~0.2s when adding 10K notes).